server.go

```diff
@@ -100,6 +106,38 @@ func (srv *server) requireAuth(w http.ResponseWriter, req *http.Request, next ht
 	jsonRespond(w, http.StatusForbidden, &jsonErr{Err: errForbidden})
 }
 
+func (srv *server) instAuthd(f http.HandlerFunc) http.Handler {
+	return negroni.New(negroni.HandlerFunc(srv.requireInstAuth), negroni.Wrap(http.HandlerFunc(f)))
+}
+
+func (srv *server) requireInstAuth(w http.ResponseWriter, req *http.Request, next http.HandlerFunc) {
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" {
+		w.Header().Set("WWW-Authenticate", "token")
+		jsonRespond(w, http.StatusUnauthorized, &jsonErr{Err: errUnauthorized})
+		return
+	}
+
+	instanceID, ok := mux.Vars(req)["instance_id"]
+	if !ok {
+		jsonRespond(w, http.StatusBadRequest, &jsonErr{Err: errNoInstanceID})
+		return
+	}
+
+	instTok, err := srv.db.fetchInstanceToken(instanceID)
+	if err != nil {
+		jsonRespond(w, http.StatusForbidden, &jsonErr{Err: errForbidden})
+		return
+	}
+
+	if authHeader == fmt.Sprintf("token %s", instTok) {
```

Comment:
Oooh, I like this for not accidentally falling through to the "permit" call :+1: